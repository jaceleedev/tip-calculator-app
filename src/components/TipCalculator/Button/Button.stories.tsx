import type { Meta, StoryObj } from '@storybook/react';
import { fn } from '@storybook/test';

import Button from './Button';

const meta = {
  component: Button,
  parameters: {
    layout: 'centered',
  },
  tags: ['autodocs'],
  argTypes: {
    label: {
      description: 'The text content of the button',
      control: 'text',
    },
    variant: {
      description: 'The visual style variant of the button',
      control: 'select',
      options: ['primary', 'secondary'],
    },
    isButtonSelected: {
      description: 'Whether the button is in a selected state',
      control: 'boolean',
    },
    onClick: {
      description: 'Function to be called when the button is clicked',
    },
    disabled: {
      description: 'Whether the button is disabled',
      control: 'boolean',
    },
  },
  args: {
    label: '15%',
    variant: 'primary',
    isButtonSelected: false,
    onClick: fn(),
  },
  decorators: [
    (Story) => (
      <div style={{ width: '150px' }}>
        <Story />
      </div>
    ),
  ],
} satisfies Meta<typeof Button>;

export default meta;

type Story = StoryObj<typeof meta>;

export const Primary: Story = {
  args: {},
};

export const PrimarySelected: Story = {
  args: {
    isButtonSelected: true,
  },
};

export const PrimaryHover: Story = {
  parameters: {
    pseudo: { hover: true },
  },
};

export const PrimaryDisabled: Story = {
  args: {
    disabled: true,
  },
};

export const Secondary: Story = {
  args: {
    label: 'RESET',
    variant: 'secondary',
  },
};

export const SecondaryHover: Story = {
  args: {
    label: 'RESET',
    variant: 'secondary',
  },
  parameters: {
    pseudo: { hover: true },
  },
};

export const SecondaryDisabled: Story = {
  args: {
    label: 'RESET',
    variant: 'secondary',
    disabled: true,
  },
};
