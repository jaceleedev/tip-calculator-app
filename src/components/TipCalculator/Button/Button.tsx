import { ButtonHTMLAttributes } from 'react';
import { clsx } from 'clsx';

interface ButtonProps extends ButtonHTMLAttributes<HTMLButtonElement> {
  label: string;
  variant: 'primary' | 'secondary';
  isButtonSelected?: boolean;
}

function Button({
  label,
  variant,
  isButtonSelected = false,
  onClick,
  ...rest
}: Readonly<ButtonProps>) {
  const baseClasses = `flex justify-center items-center w-full h-12 rounded-[5px] hover:bg-light-strong-cyan`;
  const variantClasses = {
    primary: clsx(
      'text-primary hover:text-very-dark-cyan',
      'disabled:text-white disabled:bg-very-dark-cyan',
      isButtonSelected
        ? 'text-very-dark-cyan bg-strong-cyan'
        : 'text-white bg-very-dark-cyan'
    ),
    secondary: clsx(
      'text-secondary text-very-dark-cyan bg-strong-cyan',
      'disabled:text-opacity-35 disabled:bg-very-dark-grayish-cyan'
    ),
  };
  const buttonClasses = clsx(baseClasses, variantClasses[variant]);

  return (
    <button className={buttonClasses} onClick={onClick} {...rest}>
      {label}
    </button>
  );
}

export default Button;
