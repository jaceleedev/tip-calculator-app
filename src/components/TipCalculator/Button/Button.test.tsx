import { describe, expect, it, vi } from 'vitest';
import { fireEvent, render, screen } from '@testing-library/react';
import { axe, toHaveNoViolations } from 'jest-axe';
import Button from './Button';

expect.extend(toHaveNoViolations);

describe('Button', () => {
  // 1. Rendering test
  it('renders correctly', () => {
    render(<Button label="render test" variant="primary" />);
    const button = screen.getByRole('button', { name: 'render test' });
    expect(button).toBeDefined();
  });

  // 2. Props test
  it('applies correct variant classes', () => {
    render(<Button label="primary button" variant="primary" />);
    const primaryButton = screen.getByRole('button', {
      name: 'primary button',
    });
    expect(primaryButton.className).toContain('text-white');
    expect(primaryButton.className).toContain('bg-very-dark-cyan');

    render(<Button label="secondary button" variant="secondary" />);
    const secondaryButton = screen.getByRole('button', {
      name: 'secondary button',
    });
    expect(secondaryButton.className).toContain('text-very-dark-cyan');
    expect(secondaryButton.className).toContain('bg-strong-cyan');
  });

  // 3. isButtonSelected prop test
  it('applies correct classes when selected', () => {
    render(
      <Button
        label="selected Button"
        variant="primary"
        isButtonSelected={true}
      />
    );
    const selectedButton = screen.getByRole('button', {
      name: 'selected Button',
    });
    expect(selectedButton.className).toContain('text-very-dark-cyan');
    expect(selectedButton.className).toContain('bg-strong-cyan');
  });

  // 4. Event test
  it('calls onClick when clicked', () => {
    const handleClick = vi.fn();
    render(
      <Button label="click test" variant="primary" onClick={handleClick} />
    );
    const button = screen.getByRole('button', { name: 'click test' });
    fireEvent.click(button);
    expect(handleClick).toHaveBeenCalledOnce();
  });

  // 6. Hover state test
  it('has correct hover classes', () => {
    render(<Button label="hover test" variant="primary" />);
    const button = screen.getByRole('button', { name: 'hover test' });
    fireEvent.mouseOver(button);
    expect(button.className).toContain('bg-light-strong-cyan');
    expect(button.className).toContain('text-very-dark-cyan');
  });

  // 7. Disabled state test
  it('applies correct classes when disabled', () => {
    render(<Button label="disabled test" variant="primary" disabled />);
    const button = screen.getByRole('button', { name: 'disabled test' });
    expect(button.className).toContain('text-white');
    expect(button.className).toContain('bg-very-dark-cyan');
  });

  // 8. Accessibility test
  it('has no accessibility violations', async () => {
    const { container } = render(
      <Button label="accessibility test" variant="primary" />
    );
    const results = await axe(container);
    expect(results).toHaveNoViolations();
  });

  // 9. Snapshot test
  it('matches snapshot', () => {
    const { asFragment } = render(
      <Button label="snapshot test" variant="primary" />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
