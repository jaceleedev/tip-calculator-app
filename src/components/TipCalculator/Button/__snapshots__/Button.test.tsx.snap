// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button > matches snapshot 1`] = `
<DocumentFragment>
  <button
    class="flex justify-center items-center w-full h-12 rounded-[5px] hover:bg-light-strong-cyan text-primary hover:text-very-dark-cyan disabled:text-white disabled:bg-very-dark-cyan text-white bg-very-dark-cyan"
  >
    snapshot test
  </button>
</DocumentFragment>
`;
